controller_interface/include/controller_interface/async_function_handler.hpp

```diff
@@ -0,0 +1,242 @@
+#include "rclcpp/time.hpp"
+#include "rclcpp_lifecycle/state.hpp"
+
+namespace ros2_control
+{
+/**
+ * @brief Class to handle asynchronous function calls.
+ * AsyncFunctionHandler is a class that allows the user to have a ascynchronous call to the parsed
+ * method, when the lifecycyle state is in the ACTIVE state
+ */
+template <typename T>
+class AsyncFunctionHandler
+{
+public:
+  AsyncFunctionHandler() = default;
+
+  ~AsyncFunctionHandler() { stop_async_update(); }
+
+  /// Initialize the AsyncFunctionHandler with the get_state_function and async_function
+  /**
+   * @param get_state_function Function that returns the current lifecycle state
+   * @param async_function Function that will be called asynchronously
+   * If the AsyncFunctionHandler is already initialized and is running, it will throw a runtime
+   * error.
+   * If the parsed functions are not valid, it will throw a runtime error.
+   */
+  void init(
+    std::function<const rclcpp_lifecycle::State &()> get_state_function,
+    std::function<T(const rclcpp::Time &, const rclcpp::Duration &)> async_function)
+  {
+    if (get_state_function == nullptr)
+    {
+      throw std::runtime_error(
+        "AsyncFunctionHandler: parsed function to get the lifecycle state is not valid!");
+    }
+    if (async_function == nullptr)
+    {
+      throw std::runtime_error(
+        "AsyncFunctionHandler: parsed function to call asynchronously is not valid!");
+    }
+    if (thread_.joinable())
+    {
+      throw std::runtime_error(
+        "AsyncFunctionHandler: Cannot reinitialize while the thread is "
+        "running. Please stop the async update first!");
+    }
+    get_state_function_ = get_state_function;
+    async_function_ = async_function;
+  }
+
+  /// Triggers the async update method cycle
+  /**
+   * @param time Current time
+   * @param period Current period
+   * @return A pair with the first element being a boolean indicating if the async update method was
+   * triggered and the second element being the last return value of the async function.
+   * If the AsyncFunctionHandler is not initialized properly, it will throw a runtime error.
+   * If the async update method is waiting for the trigger, it will notify the async thread to start
+   * the update cycle.
+   * If the async update method is still running, it will return the last return value of the async
+   * function.
+   *
+   * \note In the case of controllers, The controller manager is responsible
+   * for triggering and maintaining the controller's update rate, as it should be only acting as a
+   * scheduler. Same applies to the resource manager when handling the hardware components.
+   */
+  std::pair<bool, T> trigger_async_update(
+    const rclcpp::Time & time, const rclcpp::Duration & period)
+  {
+    initialize_async_update_thread();
+    std::unique_lock<std::mutex> lock(async_mtx_, std::try_to_lock);
+    bool trigger_status = false;
+    if (!trigger_in_progress_ && lock.owns_lock())
```

Comment:
I feel more comfortable with:
```suggestion
    if (lock.owns_lock() && !trigger_in_progress_)
```